### PR TITLE
Feature:make tree dynamic to add elements

### PR DESCRIPTION
**Pull request by tomip01**
* Can add dynamically new elements with `add()` method
* Refactor `new()` method  with a new method `build()` 
* Add unit tests

### CODE REVIEW

File: src/merkle.rs [hunk 1]
```diff
@@ -312,4 +386,60 @@ mod tests {
...
+            concat_hash(&first_level[2], &first_level[2]),
+        ];
+        let root = vec![concat_hash(&second_level[0], &second_level[1])];
+
+        assert_eq!(leaf_hash, merkle.tree[0]);
+        assert_eq!(first_level, merkle.tree[1]);
+        assert_eq!(second_level, merkle.tree[2]);
+        assert_eq!(root, merkle.tree[3]);
+        assert_eq!(4, merkle.tree.len());
+    }
```
Comment (LucasUTNFRD) [inline src/merkle.rs]: Well done Tomi, it will be nice to have test to coverage edge cases of adding a new element where the tree increases the height.

Comment (tomip01) [reply src/merkle.rs]: Thanks!, I think the second test you marked tests that, it changes from a 3-level tree to a 4-level tree. Maybe I didn't understand your question or the test name is wrong.

Comment (LucasUTNFRD) [reply src/merkle.rs]: Sorry Tomi, i got it wrong!

**Review by LucasUTNFRD — APPROVED** [review]

File: src/merkle.rs [hunk 2]
```diff
@@ -146,6 +155,71 @@ impl MerkleTree {
...
+            let new_hash = if actual_index % 2 == 0 {
+                concat_hash(self_hash, sibling_hash)
+            } else {
+                concat_hash(sibling_hash, self_hash)
+            };
+
+            actual_index /= 2;
+
+            // now check if element is present, then update the hash (only occur when the same hash is used to create a new one)
+            // if not present, push the new hash
```
Comment (lima-limon-inc) [inline src/merkle.rs]: Slight nitpick
```suggestion
            // now check if the element is present, then update the hash (only occurs when the same hash is used to create a new one)
            // if not present, push the new hash
```

File: src/merkle.rs [hunk 3]
```diff
@@ -146,6 +155,71 @@ impl MerkleTree {
...
+            match self.tree[i + 1].get(actual_index) {
+                Some(_) => self.tree[i + 1][actual_index] = new_hash,
+                None => self.tree[i + 1].push(new_hash),
+            }
+        }
+
+        // this is for when the tree raise one level. The previous root level now has two elements instead of one
+        // then we need to create a new level with the hash of the two elements concatenated
```
Comment (lima-limon-inc) [inline src/merkle.rs]: ditto
```suggestion
        // this is for when the tree is raised one level. The previous root level now has two elements instead of one
        // then we need to create a new level with the hash of the two elements concatenated
```

**Review by lima-limon-inc — APPROVED** [review]
LGTM, great Job Tomi ^_^
